lib/teamsnap.rb

```diff
@@ -0,0 +1,261 @@
+      ) do |c|
+        c.request :teamsnap_auth_middleware, {:token => token}
+        c.adapter :typhoeus
+      end
+
+      collection = TeamSnap.run(:get, "/")
+
+      classes = []
+      client.in_parallel do
+        classes = collection
+          .fetch(:links)
+          .map { |link| classify_rel(link) }
+          .compact
+      end
+
+      classes.each { |cls| cls.parse_collection }
+
+      apply_endpoints(self, collection) && true
+    end
+
+    def run(via, href, args = {})
+      resp = client.send(via, href, args)
+
+      if resp.status == 200
+        Oj.load(resp.body)
+          .fetch(:collection)
+      else
+        error_message = parse_error(resp)
+        raise TeamSnap::Error.new(error_message)
+      end
+    end
+
+    def parse_error(resp)
+      Oj.load(resp.body)
+        .fetch(:collection)
+        .fetch(:error)
+        .fetch(:message)
+    end
+
+    def load_items(collection)
+      collection
+        .fetch(:items) { [] }
+        .map { |item|
+          data = parse_data(item)
+          type = type_of(item)
+          cls  = load_class(type, data)
+
+          cls.new(data).tap { |obj|
+            obj.send(:load_links, item.fetch(:links))
+          }
+        }
+    end
+
+    def apply_endpoints(obj, collection)
+      collection
+        .fetch(:queries) { [] }
+        .each { |endpoint| register_endpoint(obj, endpoint, :via => :get) }
+
+      collection
+        .fetch(:commands) { [] }
+        .each { |endpoint| register_endpoint(obj, endpoint, :via => :post) }
+    end
+
+    private
+
+    attr_accessor :client
+
+    def classify_rel(link)
+      return if EXCLUDED_RELS.include?(link.fetch(:rel))
+
+      rel = link.fetch(:rel)
+      href = link.fetch(:href)
+      name = Inflecto.classify(rel)
+      resp = client.get(href)
+
+      TeamSnap.const_set(
+        name, Class.new { include TeamSnap.collection(href, resp) }
+      ) unless TeamSnap.const_defined?(name)
+    end
+
+    def register_endpoint(obj, endpoint, opts)
+      rel = endpoint.fetch(:rel)
+      href = endpoint.fetch(:href)
+      valid_args = endpoint.fetch(:data) { [] }
+        .map { |datum| datum.fetch(:name).to_sym }
+      via = opts.fetch(:via)
+
+      obj.define_singleton_method(rel) do |*args|
+        args = Hash[*args]
+
+        unless args.all? { |arg, _| valid_args.include?(arg) }
+          raise ArgumentError.new(
+            "Invalid argument(s). Valid argument(s) are #{valid_args.inspect}"
+          )
+        end
+
+        TeamSnap.load_items(
+          TeamSnap.run(via, href, args)
+        )
+      end
+    end
+
+    def parse_data(item)
+      item
+        .fetch(:data)
+        .map { |datum|
+          name = datum.fetch(:name)
+          value = datum.fetch(:value)
+          type = datum.fetch(:type) { :default }
+
+          value = DateTime.parse(value) if value && type == "DateTime"
+
+          [name, value]
+        }
+        .inject({}) { |h, (k, v)| h[k] = v; h }
```

Comment:
torn about this, I am generally not a fan of the single letter variables, even though this is a pretty common pattern. Would it put too much to remove the `map` and `inject` directly into the hash after the `fetch`?